cache/cache.cpp

```diff
@@ -190,11 +194,332 @@ size_t cache_manager::idx(const unsigned char *id) {
+static int dnet_cmd_cache_io_write(struct cache_manager *cache,
+                                   struct dnet_net_state *st,
+                                   struct dnet_cmd *cmd,
+                                   struct dnet_io_attr *io,
+                                   char *data)
+{
+	int err;
+	write_response write_resp;
+
+	std::tie(write_resp, err) = cache->write(io->id, st, cmd, write_request(io, data));
+
+	switch (write_resp) {
+		case write_response::ERROR: break;
+		case write_response::HANDLED_IN_BACKEND:
+			cmd->flags &= ~DNET_FLAGS_NEED_ACK;
+			break;
+		case write_response::HANDLED_IN_CACHE:
+			cmd->flags &= ~DNET_FLAGS_NEED_ACK;
+			err = dnet_send_file_info_ts_without_fd(st, cmd, data, io->size, &io->timestamp);
+			break;
+	}
+
+	return err;
+}
+
+static int dnet_cmd_cache_io_write_new(struct cache_manager *cache,
+                                       struct dnet_net_state *st,
+                                       struct dnet_cmd *cmd,
+                                       void *data)
+{
+	using namespace ioremap::elliptics;
+
+	auto data_p = data_pointer::from_raw(data, cmd->size);
+
+	auto request = [&data_p] () {
+		size_t offset = 0;
+		dnet_write_request request;
+		deserialize(data_p, request, offset);
+		data_p = data_p.skip(offset);
+		return request;
+	} ();
+
+	if (request.ioflags & DNET_IO_FLAGS_NOCACHE) {
+		return -ENOTSUP;
+	}
+
+	int err;
+	write_response write_resp;
+
+	std::tie(write_resp, err) = cache->write(reinterpret_cast<unsigned char *>(&cmd->id.id), st, cmd,
+						 write_request(request, data, reinterpret_cast<char *>(data_p.data())));
+
+	// TODO: return data_t from write() and fill response properly
+	auto response = serialize(dnet_lookup_response{
+			0,
+			request.user_flags,
+			"",
+
+			request.json_timestamp,
+			0,
+			request.json_size,
+			request.json_capacity,
+
+			request.timestamp,
+			0,
+			request.data_size,
+	});
+
+	switch (write_resp) {
+		case write_response::ERROR: break;
+		case write_response::HANDLED_IN_BACKEND:
+			cmd->flags &= ~DNET_FLAGS_NEED_ACK;
+			break;
+		case write_response::HANDLED_IN_CACHE:
+			cmd->flags &= ~DNET_FLAGS_NEED_ACK;
+			err = dnet_send_reply(st, cmd, response.data(), response.size(), 0);
+			break;
+	}
+
+	return err;
+}
+
+static int dnet_cmd_cache_io_read(struct cache_manager *cache,
+                                  struct dnet_net_state *st,
+                                  struct dnet_cmd *cmd,
+                                  struct dnet_io_attr *io)
+{
+	struct dnet_node *n = st->n;
+
+	auto it = cache->read(io->id, io->flags);
+	if (!it) {
+		if (!(io->flags & DNET_IO_FLAGS_CACHE)) {
+			return -ENOTSUP;
+		}
+
+		return -ENOENT;
+	}
+
+	auto d = it->data;
+
+	/*!
+	 * When offset is larger then size of the file, operation is definitely incorrect
+	 */
+	if (io->offset >= d->size()) {
+		BH_LOG(*n->log, DNET_LOG_ERROR, "%s: %s cache: invalid offset: "
```

Comment:
`BH_LOG(*n->log, ...` can be replaced by `dnet_log(st->n, ...` or `dnet_log(n, ...)`.